File: cloud-functions/submitTrade/exchanges/bingx/execution.js
const CustomError = require('../../utils/error.js');
const BingXFunctions = require('./api/perpetual.js');
const { storeTrade, storeTP, storeSL, getTradeInfo, getTpOrders, updateTradeQuantity, getTpSlOrders, getSpecificOrder } = require('../../utils/firestore.js');
const { convertSymbol, roundToPrecision } = require('./scripts/settings.js');
const sendCancel = require('./scripts/cancel.js');
const Order = require('./scripts/orderFactory.js');
const calculateTpAmounts = require('./scripts/distribution.js');
const distributionPercentages = require('../../utils/partial.js');
const { getTpsStatus } = require('./scripts/order.js');
const { v4: uuidv4 } = require('uuid');

async function bulkOrder(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, margin, trader_exchange, marginType } = data;
        const { side, leverage, entry, take_profits, stop_losses } = data.payload;
        console.log(data);
        let takeProfits = take_profits;
        let stopLosses = stop_losses;

        let symbol = data.payload.symbol;

        let orderType = entry != "market" ? "LIMIT" : "MARKET";

        // Convert the symbol to the format needed for bingx
        symbol = await convertSymbol(symbol);

        const getPrecisionPromise = session.getPrecisions(symbol);
        const switchMarginModePromise = session.switchMarginMode(symbol, marginType);
        const setLeveragePromise = session.setLeverage(symbol, side, leverage);
    
        const [precision] = await Promise.all([getPrecisionPromise, switchMarginModePromise, setLeveragePromise]);

        // Calculate the quantity for the main order depends on if market or limit order
        let quantity;
        if (entry !== "market") {
          quantity = roundToPrecision((parseFloat(margin) * parseInt(leverage) / parseFloat(entry)), precision.quantityPrecision);
        } else {
          const fetchPrice = await session.getMarket(symbol)
          const marketPrice = fetchPrice["lastPrice"];
          quantity = roundToPrecision((parseFloat(margin) * parseInt(leverage) / parseFloat(marketPrice)), precision.quantityPrecision);
        }

        let preparedOrders = [];

        let orderId = String(uuidv4())
        preparedOrders.push(new Order(symbol, orderType, side.toUpperCase(), entry === "market" ? null : entry, quantity, side === "Buy" ? "LONG" : "SHORT", null, orderId))

        let newTakeProfits = await calculateTpAmounts(takeProfits, quantity, precision);

        let tpSlPositionSide = side === "Buy" ? "LONG" : "SHORT"

        let tpPrice;
        newTakeProfits.forEach(tp => {
            tp.orderId = String(uuidv4());
            tp.tradeId = tradeId;
            tpPrice = roundToPrecision(parseFloat(tp.tp_value), precision.pricePrecision);
            preparedOrders.push(new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", null, tp.tp_amount, tpSlPositionSide, tpPrice, tp.orderId));
        });
        
        let slPrice;
        stopLosses.forEach(sl => {
            sl.orderId = String(uuidv4());
            sl.tradeId = tradeId;
            slPrice = roundToPrecision(parseFloat(sl.sl_value), precision.pricePrecision);
            sl.sl_amount = roundToPrecision(parseFloat(quantity) * parseFloat(sl.sl_percentage), precision.quantityPrecision)
            preparedOrders.push(new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", null, sl.sl_amount, tpSlPositionSide, slPrice, sl.orderId));
        });

        await Promise.all(preparedOrders.map(order => session.tradeOrder(order)));

        // Store trade info
        const tradeInfo = {
            tradeId,
            orderId,
            symbol,
            "type": orderType,
            side,
            quantity,
            entry,
            leverage,
            margin,
            "exchange": trader_exchange
        }

        // Storing trade/takeprofits/stoplosses concurrently
        const tradePromise = storeTrade(traderId, tradeInfo)
        const tpPromises = newTakeProfits.map(tp => storeTP(traderId, tp));
        const slPromises = stopLosses.map(sl => storeSL(traderId, sl));
        await Promise.all([tradePromise, ...tpPromises, ...slPromises]);

        return

    } catch(error) {
        throw new CustomError({
            message: `Error handling the trade in execution: ${error.message}`,
            status: 500,
            source: 'bulkOrder',
        });
    }
}

async function sendSl(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, document_id, payload } = data;

        const tradeInfo = await getTradeInfo(traderId, tradeId);
        const precision = await session.getPrecisions(symbol);

        let side = tradeInfo.side;
        let slPositionSide = side.toUpperCase() === "BUY" ? "LONG" : "SHORT"

        let position = await session.getPosition(symbol);

        let positionQuantity;
        if (position.length !== 0) {
            positionQuantity = Math.abs(parseFloat(position[0].positionAmt))
        } else {
            positionQuantity = tradeInfo.quantity;
        }

        let orderId = String(uuidv4());

        let price = roundToPrecision(parseFloat(payload.sl_value), precision.pricePrecision);

        let order = new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", price, positionQuantity, slPositionSide, price, orderId)

        await session.tradeOrder(order);

        payload.orderId = orderId;
        payload.sl_amount = positionQuantity;
        await storeSL(traderId, payload)

        return

    } catch(error) {
        throw new CustomError({
            message: `Error submitting SL in execution: ${error.message}`,
            status: 500,
            source: 'sendSl',
        });
    }
}

async function replaceSl(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, document_id, payload } = data;

        const tradeInfo = await getTradeInfo(traderId, tradeId);

        let side = tradeInfo.side;
        let slPositionSide = side.toUpperCase() === "BUY" ? "LONG" : "SHORT"

        // Cancel the current stoploss
        await sendCancel(session, tradeInfo.symbol, traderId, tradeId, document_id, "sl")      

        let position = await session.getPosition(symbol);

        let positionQuantity;
        if (position.length !== 0) {
            positionQuantity = Math.abs(parseFloat(position[0].positionAmt))
        } else {
            positionQuantity = tradeInfo.quantity;
        }

        let orderId = String(uuidv4());
        let price = roundToPrecision(parseFloat(payload.sl_value), precision.pricePrecision);

        let order = new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", price, positionQuantity, slPositionSide, price, orderId)

        await session.tradeOrder(order);

        payload.orderId = orderId;
        payload.sl_amount = positionQuantity;
        await storeSL(traderId, payload)

        return

    } catch(error) {
        throw new CustomError({
            message: `Error replacing SL in execution: ${error.message}`,
            status: 500,
            source: 'replaceSl',
        });
    }
}

async function cancelOrder(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, document_id, trade_type } = data;

        const tradeInfo = await getTradeInfo(traderId, tradeId);
        let orders = await session.currentOrders(tradeInfo.symbol);
        console.log(BigInt(orders.orders[0].orderId));
        await session.getOrder(tradeInfo.symbol, BigInt(orders.orders[0].orderId), null);
        await sendCancel(session, tradeInfo.symbol, traderId, tradeId, document_id, trade_type)   
        
        return

    } catch(error) {
        throw new CustomError({
            message: `Error cancelling order in execution: ${error.message}`,
            status: 500,
            source: 'cancelOrder',
        });
    }
}

async function cancelAllOrders(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId } = data;

        const tradeInfo = await getTradeInfo(traderId, tradeId);

        const symbol = tradeInfo.symbol;

        let position = await session.getPosition(symbol)    

        let quantity, positionSide;
        if (position.length !== 0) {
            quantity = position[0]["positionAmt"]
            positionSide = position[0]["positionSide"]

            let emergencySide = positionSide == "LONG" ? "SELL": "BUY"
            let orderId = String(uuidv4())

            let emergencyOrder = new Order(symbol, "MARKET", emergencySide, null, quantity, positionSide, null, orderId);

            emergency = await session.tradeOrder(emergencyOrder);

        } else {
            let orderId = tradeInfo.orderID
            await session.cancelOrder(symbol, null, orderId)
        }

        await session.cancelAllOrders(symbol);

        return

    } catch(error) {
        throw new CustomError({
            message: `Error cancelling all orders in execution: ${error.message}`,
            status: 500,
            source: 'cancelAllOrders',
        });
    }
}

async function cancelAllTps(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId } = data;

        const tradeInfoPromise = getTradeInfo(traderId, tradeId);
        const tpOrdersPromise = getTpOrders(traderId, tradeId);

        const [tradeInfo, tpOrders] = await Promise.all([tradeInfoPromise, tpOrdersPromise]);
        
        let symbol = tradeInfo.symbol;

        await Promise.all(tpOrders.map(order => session.cancelOrder(symbol, null, order.orderID)));

        return

    } catch(error) {
        throw new CustomError({
            message: `Error cancelling all tps in execution: ${error.message}`,
            status: 500,
            source: 'cancelAllTps',
        });
    }
}

async function bulkTp(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, take_profits } = data;

        const tradeInfo = await getTradeInfo(traderId, tradeId);
        const symbol = tradeInfo.symbol;

        let takeProfits = take_profits;

        const precision = await session.getPrecisions(symbol);

        let newTakeProfits = await calculateTpAmounts(takeProfits, quantity, precision);

        let preparedOrders = [];

        let side = tradeInfo.side;
        let tpPositionSide = side === "BUY" ? "LONG" : "SHORT"

        let tpPrice;
        newTakeProfits.forEach(tp => {
            tp.orderId = String(uuidv4());
            tpPrice = roundToPrecision(parseFloat(tp.tp_value), precision.pricePrecision);
            preparedOrders.push(new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", tpPrice, tp.tp_amount, tpPositionSide, tpPrice, tp.orderId));
        });

        await Promise.all(preparedOrders.map(order => session.tradeOrder(order)));

        const tpPromises = newTakeProfits.map(tp => storeTP(traderId, tp));

        await Promise.all([...tpPromises]);

        return

    } catch(error) {
        throw new CustomError({
            message: `Error handling takeprofits in execution: ${error.message}`,
            status: 500,
            source: 'bulkTp',
        });
    }
}

async function partialClose(apiKey, apiSecret, data) {
    try {
        const session = new BingXFunctions(apiKey, apiSecret);
        const { traderId, tradeId, percentage } = data;

        const tradeInfoPromise = getTradeInfo(traderId, tradeId);
        const TpSlOrdersPromise = getTpSlOrders(traderId, tradeId);

        const [tradeInfo, TpSlOrders] = await Promise.all([tradeInfoPromise, TpSlOrdersPromise]);

        let tpOrders = TpSlOrders.filter(order => order.tradeType === 'tp');
        let slOrders = TpSlOrders.filter(order => order.tradeType === 'sl');

        let side = tradeInfo.side;
        let tpSlPositionSide = side.toUpperCase() === "BUY" ? "LONG" : "SHORT"

        let symbol = tradeInfo.symbol;

        const precisionPromise = session.getPrecisions(symbol);
        const positionPromise = session.getPosition(symbol);

        const [precision, position] = await Promise.all([precisionPromise, positionPromise]);

        let positionQuantity;
        let executed;
        if (position.length !== 0) {
            executed = true
            positionQuantity = Math.abs(parseFloat(position[0].positionAmt))
        } else {
            executed = false
            positionQuantity = tradeInfo.quantity;
        }

        const quantityToSell = parseFloat((parseFloat(positionQuantity) * percentage).toFixed(precision.quantityPrecision));

        let tpsData = await getTpsStatus(session, symbol, tpOrders);
        let takeProfits = distributionPercentages(tpsData);

        let newTakeProfits = await calculateTpAmounts(takeProfits, newQuantity, precision)

        await session.cancelAllOrders(symbol);

        let preparedOrders = [];

        if (executed) {
            let sellOrder = new Order(symbol, "MARKET", side == "SELL" ? "BUY" : "SELL", null, positionQuantity, tpSlPositionSide, null, null);
            await session.tradeOrder(sellOrder);
            await updateTradeQuantity(traderId, tradeId, newQuantity);

        } else {
            const newQuantity = parseFloat((parseFloat(positionQuantity) - quantityToSell).toFixed(precision.quantityPrecision));

            let orderId = String(uuidv4()); 
            let order = new Order(symbol, "LIMIT", side.toUpperCase(), tradeInfo["entry"], newQuantity, null, null, orderId)
            
            await session.tradeOrder(order);
            await updateTradeQuantity(traderId, tradeId, newQuantity);
        }

        let tpPrice;
        newTakeProfits.forEach(tp => {
            tp.orderId = String(uuidv4());
            tpPrice = roundToPrecision(parseFloat(tp.tp_value), precision.pricePrecision);
            preparedOrders.push(new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", tpPrice, tp.tp_amount, tpSlPositionSide, tpPrice, tp.orderId));
        });

        let slPrice;
        slOrders.forEach(sl => {
            sl.orderId = String(uuidv4());
            slPrice = roundToPrecision(parseFloat(sl.sl_value), precision.pricePrecision);
            preparedOrders.push(new Order(symbol, "TRIGGER_MARKET", side === "Buy" ? "SELL" : "BUY", slPrice, sl.sl_amount, tpSlPositionSide, slPrice, sl.orderId));
        });

        await Promise.all(preparedOrders.map(order => session.tradeOrder(order)));

        const tpPromises = newTakeProfits.map(tp => storeTP(traderId, tp));
        const slPromises = slOrders.map(sl => storeSL(traderId, sl));
        await Promise.all([...tpPromises, ...slPromises]);

        return

    } catch(error) {
        throw new CustomError({
            message: `Error partially closing trade in execution: ${error.message}`,
            status: 500,
            source: 'partialClose',
        });
    }
}


module.exports = {
    bulkOrder,
    cancelAllOrders,
    replaceSl,
    cancelAllTps,
    bulkTp,
    partialClose,
    sendSl,
    cancelOrder
};

File: cloud-functions/submitTrade-py/exchanges/binance/execution.py
import asyncio
import logging
from .api.perpetual import BinanceFunctions
from utils.firestore import store_trade, store_tp, store_sl, get_trade_info, update_trade_quantity, get_tp_sl_orders, \
    get_tp_orders, delete_tp_sl_order
from utils.partial import distribute_percentages
from .scripts.order_factory import Order
from .scripts.cancel import send_cancel
from .scripts.settings import get_position_quantity
from .scripts.distribution import calculate_tp_amounts
from .scripts.order import get_tps_status

logger = logging.getLogger(__name__)


async def bulk_order(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        margin = data['margin']
        trader_exchange = data['trader_exchange']
        margin_type = data['margin_type']

        symbol = data['payload']['symbol']
        side = data['payload']['side'].capitalize()
        leverage = data['payload']['leverage']
        entry = data['payload']['entry']
        take_profits = data['payload']['take_profits']
        stop_losses = data['payload']['stop_losses']

        order_type = "LIMIT" if entry != "market" else "MARKET"

        precision, set_leverage, margin_mode = await asyncio.gather(
            session.get_precisions(symbol),
            session.set_leverage(symbol, leverage),
            session.switch_margin_mode(symbol, margin_type)
        )

        market_price = float(await session.get_market(symbol))

        quantity = round((float(margin) * int(leverage) / float(entry)),
                         precision["quantity_precision"]) if entry != "market" else round(
            (float(margin) * int(leverage) / market_price), precision["quantity_precision"])

        initial_order = Order(symbol, order_type, side, entry, quantity, None, False)

        prepared_orders = [initial_order]

        new_take_profits = await calculate_tp_amounts(take_profits, quantity, precision)

        tp_sl_side = "SELL" if side == "BUY" else "BUY"

        for tp in new_take_profits:
            tp_price = round(float(tp['tp_value']), precision["price_precision"])
            tp_order = Order(symbol, "TAKE_PROFIT_MARKET", tp_sl_side, None, tp['tp_amount'], tp_price, True)
            prepared_orders.append(tp_order)

        for sl in stop_losses:
            sl_price = round(float(sl['sl_value']), precision["price_precision"])
            sl['sl_amount'] = round(float(quantity) * float(sl['sl_percentage']), precision["quantity_precision"])
            sl_order = Order(symbol, "STOP_MARKET", tp_sl_side, None, sl['sl_amount'], sl_price, True)
            prepared_orders.append(sl_order)

        order_ids = await asyncio.gather(*(session.trade_order(order) for order in prepared_orders))

        new_take_profits_with_ids = []
        stop_losses_with_ids = []

        # Assign orderIds to take profits and stop losses
        tp_count = 0
        for i, tp_order in enumerate(prepared_orders[1:len(new_take_profits) + 1], start=1):
            tp_order_dict = vars(tp_order)
            tp_order_dict['order_id'] = order_ids[i]['orderId']
            tp_order_dict['tp_document_id'] = new_take_profits[tp_count]['tp_id']
            tp_order_dict['tp_number'] = new_take_profits[tp_count]['tp_number']
            tp_order_dict['tp_percentage'] = new_take_profits[tp_count]['tp_percentage']
            tp_order_dict['tp_value'] = new_take_profits[tp_count]['tp_value']
            tp_order_dict['tp_amount'] = new_take_profits[tp_count]['tp_amount']
            tp_order_dict['trade_id'] = tradeId
            new_take_profits_with_ids.append(tp_order_dict)
            tp_count += 1

        sl_count = 0
        for i, sl_order in enumerate(prepared_orders[len(new_take_profits) + 1:], start=len(new_take_profits) + 1):
            sl_order_dict = vars(sl_order)
            sl_order_dict['order_id'] = order_ids[i]['orderId']
            sl_order_dict['sl_document_id'] = stop_losses[sl_count]['sl_id']
            sl_order_dict['sl_number'] = stop_losses[sl_count]['sl_number']
            sl_order_dict['sl_percentage'] = stop_losses[sl_count]['sl_percentage']
            sl_order_dict['sl_value'] = stop_losses[sl_count]['sl_value']
            sl_order_dict['sl_amount'] = stop_losses[sl_count]['sl_amount']
            sl_order_dict['trade_id'] = tradeId
            stop_losses_with_ids.append(sl_order_dict)
            sl_count += 1

        trade_info = {
            "trade_id": tradeId,
            "order_id": order_ids[0]["orderId"],
            "symbol": symbol,
            "type": order_type,
            "side": side,
            "quantity": quantity,
            "entry": entry,
            "leverage": leverage,
            "margin": margin,
            "exchange": trader_exchange
        }

        await store_trade(traderId, trade_info)

        tp_promises = [store_tp(traderId, tp) for tp in new_take_profits_with_ids]
        sl_promises = [store_sl(traderId, sl) for sl in stop_losses_with_ids]

        await asyncio.gather(*tp_promises, *sl_promises)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def send_sl(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        document_id = data['sl_id']
        payload = data['payload']

        trade_info = await get_trade_info(traderId, tradeId)

        symbol = trade_info["symbol"]
        side = trade_info["side"]
        sl_side = "SELL" if side == "BUY" else "BUY"

        precision, position = await asyncio.gather(
            session.get_precisions(symbol),
            session.get_position(symbol)
        )

        position_quantity = get_position_quantity(position, trade_info)

        price = round(float(payload["sl_value"]), precision["price_precision"])

        order = Order(symbol, "STOP_MARKET", sl_side, None, position_quantity, price, True)

        create_order = await session.trade_order(order)

        payload['trade_id'] = tradeId
        payload["sl_amount"] = position_quantity
        payload["order_id"] = create_order["orderId"]
        payload['sl_document_id'] = document_id

        await store_sl(traderId, payload)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def replace_sl(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        document_id = data['document_id']
        payload = data['payload']

        trade_info = await get_trade_info(traderId, tradeId)

        symbol = trade_info["symbol"]
        side = trade_info["side"]
        sl_side = "BUY" if side == "SELL" else "SELL"

        precision, position, cancel = await asyncio.gather(
            session.get_precisions(symbol),
            session.get_position(symbol),
            send_cancel(session, symbol, traderId, tradeId, document_id, "sl")
        )

        position_quantity = get_position_quantity(position, trade_info)

        price = round(float(payload["sl_value"]), precision["price_precision"])

        order = Order(symbol, "STOP_MARKET", sl_side, None, position_quantity, price, True)

        create_order = await session.trade_order(order)

        payload['trade_id'] = tradeId
        payload["sl_amount"] = position_quantity
        payload["order_id"] = create_order["orderId"]
        payload['sl_document_id'] = document_id

        await store_sl(traderId, payload)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def cancel_order(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        document_id = data['document_id']
        trade_type = data['trade_type']

        trade_info = await get_trade_info(traderId, tradeId)

        symbol = trade_info["symbol"]

        await send_cancel(session, symbol, traderId, tradeId, document_id, trade_type)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def cancel_all_orders(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']

        trade_info = await get_trade_info(traderId, tradeId)

        symbol = trade_info["symbol"]

        position = await session.get_position(symbol)

        quantity = get_position_quantity(position, trade_info)

        if quantity != 0:
            side = trade_info['side']

            order = Order(symbol, "MARKET", "SELL" if side == "BUY" else "BUY", None, None, True)
            await session.trade_order(order)
        else:
            await session.cancel_order(symbol, trade_info['orderID'], None)

        await session.cancel_all_orders(symbol)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def cancel_all_tps(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']

        trade_info, tp_orders = await asyncio.gather(
            get_trade_info(traderId, tradeId),
            get_tp_orders(traderId, tradeId)
        )

        await asyncio.gather(
            *[session.cancel_order(trade_info["symbol"], order['orderID'], None) for order in tp_orders])
        await asyncio.gather(
            *[delete_tp_sl_order(traderId, tradeId, order['document_id'], "tp") for order in tp_orders])
        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def bulk_tp(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        take_profits = data['take_profits']

        trade_info = await get_trade_info(traderId, tradeId)

        symbol = trade_info["symbol"]
        tp_side = "BUY" if trade_info['side'] == "SELL" else "SELL"

        position, precision = await asyncio.gather(
            session.get_position(symbol),
            session.get_precisions(symbol)
        )

        position_quantity = get_position_quantity(position, trade_info)

        new_take_profits = await calculate_tp_amounts(take_profits, position_quantity, precision)

        prepared_orders = []

        for tp in new_take_profits:
            tp_price = round(float(tp['tp_value']), precision["price_precision"])
            tp_order = Order(symbol, "TAKE_PROFIT_MARKET", tp_side, None, tp['tp_amount'], tp_price, True)
            prepared_orders.append(tp_order)

        order_ids = await asyncio.gather(*(session.trade_order(order) for order in prepared_orders))

        new_take_profits_with_ids = []

        tp_count = 0
        # Assign orderIds to take profits
        for i, tp_order in enumerate(prepared_orders[0:len(new_take_profits)], start=0):
            tp_order_dict = vars(tp_order)
            tp_order_dict['order_id'] = order_ids[i]['orderId']
            tp_order_dict['tp_document_id'] = new_take_profits[tp_count]['tp_id']
            tp_order_dict['tp_number'] = new_take_profits[tp_count]['tp_number']
            tp_order_dict['tp_percentage'] = new_take_profits[tp_count]['tp_percentage']
            tp_order_dict['tp_value'] = new_take_profits[tp_count]['tp_value']
            tp_order_dict['tp_amount'] = new_take_profits[tp_count]['tp_amount']
            tp_order_dict['trade_id'] = tradeId
            new_take_profits_with_ids.append(tp_order_dict)
            tp_count += 1

        # Store take profits in firestore
        tp_promises = [store_tp(traderId, tp) for tp in new_take_profits_with_ids]

        await asyncio.gather(*tp_promises)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)


async def partial_close(api_key, api_secret, data):
    try:
        session = BinanceFunctions(api_key, api_secret)

        traderId = data['traderId']
        tradeId = data['tradeId']
        percentage = data['percentage']

        trade_info, tp_sl_orders = await asyncio.gather(
            get_trade_info(traderId, tradeId),
            get_tp_sl_orders(traderId, tradeId)
        )

        symbol = trade_info["symbol"]
        side = trade_info["side"]
        tp_sl_side = "BUY" if side == "SELL" else "SELL"

        # Split tp_sl_orders into tp/sl orders
        tp_orders = [order for order in tp_sl_orders if order['trade_type'] == 'tp']
        sl_orders = [order for order in tp_sl_orders if order['trade_type'] == 'sl']

        # Fetch the current position and precisions
        position, precision = await asyncio.gather(
            session.get_position(symbol),
            session.get_precisions(symbol)
        )

        executed = True if float(position['positionAmt']) != 0 else False

        position_quantity = get_position_quantity(position, trade_info)

        quantity_to_sell = round(position_quantity * float(percentage), precision["quantity_precision"])
        new_quantity = round(position_quantity - quantity_to_sell, precision["quantity_precision"])

        tps_data = await get_tps_status(session, tp_orders, trade_info)

        take_profits = distribute_percentages(tps_data)

        new_take_profits = await calculate_tp_amounts(take_profits, new_quantity, precision)

        await session.cancel_all_orders(symbol)

        if executed:
            sell_order = Order(symbol, "MARKET", tp_sl_side, None, quantity_to_sell, None, True)
            await session.trade_order(sell_order)
            await update_trade_quantity(traderId, tradeId, new_quantity)
        else:
            order = Order(symbol, "LIMIT", tp_sl_side, trade_info['entry'], new_quantity, None, False)

            create_order = await session.trade_order(order)

            trade_info = {
                "trade_id": tradeId,
                "order_id": create_order["orderId"],
                "symbol": symbol,
                "type": "Limit",
                "side": side,
                "quantity": new_quantity,
                "entry": trade_info["entry"],
                "leverage": trade_info["leverage"],
                "margin": round(float(trade_info["margin"] * float(percentage)), 2),
                "exchange": trade_info["exchange"]
            }

            await store_trade(traderId, trade_info)

        prepared_orders = []
        new_take_profits_with_ids = []
        stop_losses_with_ids = []

        for tp in new_take_profits:
            tp_price = round(float(tp['tp_value']), precision["price_precision"])
            tp_order = Order(symbol, "TAKE_PROFIT_MARKET", tp_sl_side, None, tp['tp_amount'], tp_price, True)
            prepared_orders.append(tp_order)

        for sl in sl_orders:
            sl_price = round(float(sl['sl_value']), precision["price_precision"])
            sl['sl_amount'] = round(float(new_quantity) * float(sl['sl_percentage']), precision["quantity_precision"])
            sl_order = Order(symbol, "STOP_MARKET", tp_sl_side, None, sl['sl_amount'], sl_price, True)
            prepared_orders.append(sl_order)

        order_ids = await asyncio.gather(*(session.trade_order(order) for order in prepared_orders))

        # Assign orderIds to take profits and stop losses
        tp_count = 0
        for i, tp_order in enumerate(prepared_orders[1:len(new_take_profits) + 1], start=1):
            tp_order_dict = vars(tp_order)
            tp_order_dict['order_id'] = order_ids[i]['orderId']
            tp_order_dict['tp_document_id'] = new_take_profits[tp_count]['tp_id']
            tp_order_dict['tp_number'] = new_take_profits[tp_count]['tp_number']
            tp_order_dict['tp_percentage'] = new_take_profits[tp_count]['tp_percentage']
            tp_order_dict['tp_value'] = new_take_profits[tp_count]['tp_value']
            tp_order_dict['tp_amount'] = new_take_profits[tp_count]['tp_amount']
            tp_order_dict['trade_id'] = tradeId
            new_take_profits_with_ids.append(tp_order_dict)
            tp_count += 1

        sl_count = 0
        for i, sl_order in enumerate(prepared_orders[len(new_take_profits) + 1:], start=len(new_take_profits) + 1):
            sl_order_dict = vars(sl_order)
            sl_order_dict['order_id'] = order_ids[i]['orderId']
            sl_order_dict['sl_document_id'] = sl_orders[sl_count]['sl_id']
            sl_order_dict['sl_number'] = sl_orders[sl_count]['sl_number']
            sl_order_dict['sl_percentage'] = sl_orders[sl_count]['sl_percentage']
            sl_order_dict['sl_value'] = sl_orders[sl_count]['sl_value']
            sl_order_dict['sl_amount'] = sl_orders[sl_count]['sl_amount']
            sl_order_dict['trade_id'] = tradeId
            stop_losses_with_ids.append(sl_order_dict)
            sl_count += 1

        tp_promises = [store_tp(traderId, tp) for tp in new_take_profits_with_ids]
        sl_promises = [store_sl(traderId, sl) for sl in stop_losses_with_ids]

        await asyncio.gather(*tp_promises, *sl_promises)

        return

    except Exception as e:
        logger.error("An error occurred: %s", e, exc_info=True)
